src/include/btree_inline.h

```diff
@@ -1581,63 +1581,35 @@ __wt_ref_block_free(WT_SESSION_IMPL *session, WT_REF *ref)
-     * the on-disk format doesn't have space for the extra "I'm prepared" bit, so we avoid seeing
-     * them in reconciliation. Similarly, we can't skip over a page just because a transaction has
-     * deleted it and prepared; only committed transactions are suitable.)
-     *
-     * In all cases, the ref owning the page_deleted structure should be locked and its pre-lock
-     * state should be WT_REF_DELETED. This prevents the page from being instantiated while we look
-     * at it, and locks out other operations that might simultaneously discard the structure (either
-     * after checking visibility, or because its transaction aborted).
-     */
-
-    /* If the page delete info is NULL, the deletion was previously found to be globally visible. */
-    if (page_del == NULL)
-        return (true);
-
-    /* We discard page_del on transaction abort, so should never see an aborted one. */
-    WT_ASSERT(session, page_del->txnid != WT_TXN_ABORTED);
-
-    if (hide_prepared) {
-        WT_ORDERED_READ(prepare_state, page_del->prepare_state);
-        if (prepare_state == WT_PREPARE_INPROGRESS || prepare_state == WT_PREPARE_LOCKED)
-            return (false);
-    }
-
-    return (__wt_txn_visible_all(session, page_del->txnid, page_del->durable_timestamp));
-}
-
 /*
  * __wt_page_del_visible --
- *     Return if a truncate operation is visible to the caller. The same considerations apply as in
- *     the visible_all version.
+ *     Return if a truncate operation is visible. Check can be for either global visibility or for
+ *     just caller, which could be controlled through check_global_visibility.
  */
 static inline bool
-__wt_page_del_visible(WT_SESSION_IMPL *session, WT_PAGE_DELETED *page_del, bool hide_prepared)
+__wt_page_del_visible(WT_SESSION_IMPL *session, WT_PAGE_DELETED *page_del, bool hide_prepared,
+  bool check_global_visibility)
 {
     uint8_t prepare_state;
 
-    /* If the page delete info is NULL, the deletion was previously found to be globally visible. */
     if (page_del == NULL)
         return (true);
 
+    /*
+     * Check whether truncate transaction is committed or not. In case of prepared transaction,
+     * truncate operation is committed in two phases. In the first phase prepared state is set to
+     * resolved and in the later phase the committed flag of all fast truncated pages is set to
+     * true. The two phase design is to handle any page split cases, when a fast truncated page is
+     * read back and later that page split, in which case all the new split pages will not have
+     * page_del info.
+     *
+     * There is a window between setting the prepared state to resolved and setting the committed
+     * flag to true. Hence if committed flag is not set, consider it as not committed irrespective
+     * of visibility check, to avoid checking visibility when transaction commit is in progress.
+     */
+    if (!page_del->committed)
```

Comment:
Was this the entirety of the fix in the end? Initially, this PR had a change to move when the committed flag was set, but I can't see it any longer (which is good, because I don't think that was a safe change).

Reply:
Yes, this is the entirety of the fix. No change in where the committed flag is set, only an additional check of the committed flag here.